File: backend/src/main/java/com/cts/regreportx/service/TemplateService.java
package com.cts.regreportx.service;

import com.cts.regreportx.model.RegTemplate;
import com.cts.regreportx.model.TemplateField;
import com.cts.regreportx.repository.RegTemplateRepository;
import com.cts.regreportx.repository.TemplateFieldRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class TemplateService {

    private final RegTemplateRepository templateRepository;
    private final TemplateFieldRepository fieldRepository;

    @Autowired
    public TemplateService(RegTemplateRepository templateRepository, TemplateFieldRepository fieldRepository) {
        this.templateRepository = templateRepository;
        this.fieldRepository = fieldRepository;
    }


    public RegTemplate createTemplate(RegTemplate template) {
        return templateRepository.save(template);
    }

    public List<RegTemplate> getAllTemplates() {
        return templateRepository.findAll();
    }

    public Optional<RegTemplate> getTemplateById(Integer id) {
        return templateRepository.findById(id);
    }

    public RegTemplate updateTemplate(Integer id, RegTemplate templateDetails) {
        return templateRepository.findById(id).map(temp -> {
            temp.setRegulationCode(templateDetails.getRegulationCode());
            temp.setDescription(templateDetails.getDescription());
            temp.setFrequency(templateDetails.getFrequency());
            temp.setStatus(templateDetails.getStatus());
            return templateRepository.save(temp);
        }).orElseThrow(() -> new RuntimeException("Template not found with id " + id));
    }

    public void deleteTemplate(Integer id) {
        templateRepository.deleteById(id);
    }


    public TemplateField addFieldToTemplate(Integer templateId, TemplateField field) {
        field.setTemplate(templateRepository.getReferenceById(templateId));
        return fieldRepository.save(field);
    }

    public List<TemplateField> getFieldsByTemplateId(Integer templateId) {
        return fieldRepository.findByTemplate_TemplateId(templateId);
    }

    public TemplateField updateField(Integer fieldId, TemplateField fieldDetails) {
        return fieldRepository.findById(fieldId).map(field -> {
            field.setFieldName(fieldDetails.getFieldName());
            field.setDataType(fieldDetails.getDataType());
            field.setMappingExpression(fieldDetails.getMappingExpression());
            field.setRequiredFlag(fieldDetails.getRequiredFlag());
            return fieldRepository.save(field);
        }).orElseThrow(() -> new RuntimeException("Field not found with id " + fieldId));
    }

    public void deleteField(Integer fieldId) {
        fieldRepository.deleteById(fieldId);
    }
}

File: frontend/src/app/services/template.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { RegTemplate, TemplateField } from '../models/template.model';

@Injectable({
  providedIn: 'root'
})
export class TemplateService {
  private readonly API_URL = 'http://localhost:9090/api/templates';

  constructor(private http: HttpClient) {}

  getAllTemplates(): Observable<RegTemplate[]> {
    return this.http.get<RegTemplate[]>(this.API_URL);
  }

  getTemplateById(id: number): Observable<RegTemplate> {
    return this.http.get<RegTemplate>(`${this.API_URL}/${id}`);
  }

  createTemplate(template: RegTemplate): Observable<RegTemplate> {
    return this.http.post<RegTemplate>(this.API_URL, template);
  }

  updateTemplate(id: number, template: RegTemplate): Observable<RegTemplate> {
    return this.http.put<RegTemplate>(`${this.API_URL}/${id}`, template);
  }

  deleteTemplate(id: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${id}`);
  }

  getFieldsByTemplateId(templateId: number): Observable<TemplateField[]> {
    return this.http.get<TemplateField[]>(`${this.API_URL}/${templateId}/fields`);
  }

  addFieldToTemplate(templateId: number, field: TemplateField): Observable<TemplateField> {
    return this.http.post<TemplateField>(`${this.API_URL}/${templateId}/fields`, field);
  }

  updateField(fieldId: number, field: TemplateField): Observable<TemplateField> {
    return this.http.put<TemplateField>(`${this.API_URL}/fields/${fieldId}`, field);
  }

  deleteField(fieldId: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/fields/${fieldId}`);
  }
}
